File: app/src/main/java/wxm/KeepAccount/Base/db/PayNoteItem.java
package wxm.KeepAccount.Base.db;

import android.os.Parcel;
import android.os.Parcelable;

import com.j256.ormlite.field.DataType;
import com.j256.ormlite.field.DatabaseField;
import com.j256.ormlite.table.DatabaseTable;

import java.math.BigDecimal;
import java.sql.Timestamp;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import wxm.KeepAccount.Base.data.AppGobalDef;
import wxm.KeepAccount.Base.data.AppModel;

/**
 * 支出记录数据
 * Created by 123 on 2016/5/3.
 */
@DatabaseTable(tableName = "tbPayNote")
public class PayNoteItem implements Parcelable {
    public final static String FIELD_TS         = "ts";
    public final static String FIELD_USR        = "usr_id";
    public final static String FIELD_BUDGET     = "budget_id";

    @DatabaseField(generatedId = true, columnName = "_id", dataType = DataType.INTEGER)
    private int _id;

    @DatabaseField(columnName = "usr_id", foreign = true, foreignColumnName = UsrItem.FIELD_ID,
            canBeNull = false)
    private UsrItem usr;

    @DatabaseField(columnName = "budget_id", foreign = true, foreignColumnName = BudgetItem.FIELD_ID)
    private BudgetItem budget;

    @DatabaseField(columnName = "info", canBeNull = false, dataType = DataType.STRING)
    private String info;

    @DatabaseField(columnName = "note", dataType = DataType.STRING)
    private String note;

    @DatabaseField(columnName = "val", dataType = DataType.BIG_DECIMAL)
    private BigDecimal val;

    @DatabaseField(columnName = "ts", dataType = DataType.TIME_STAMP)
    private Timestamp ts;

    public int getId() {
        return _id;
    }

    public void setId(int _id) {
        this._id = _id;
    }

    public String getInfo() {
        return info;
    }

    public void setInfo(String info) {
        this.info = info;
    }

    public String getNote() {
        return note;
    }

    public void setNote(String note) {
        this.note = note;
    }

    public BigDecimal getVal() {
        return val;
    }

    public void setVal(BigDecimal val) {
        this.val = val;
    }

    public Timestamp getTs() {
        return ts;
    }

    public void setTs(Timestamp tsval) {
        this.ts = tsval;
    }

    public UsrItem getUsr() {
        return usr;
    }

    public void setUsr(UsrItem usr) {
        this.usr = usr;
    }


    public PayNoteItem()
    {
        setTs(new Timestamp(0));
        setVal(BigDecimal.ZERO);
        setInfo("");
        setNote("");

        setId(AppGobalDef.INVALID_ID);
    }

    @Override
    public String toString()
    {
        return String.format(Locale.CHINA
                ,"info : %s, val : %f, timestamp : %s\nnote : %s"
                ,getInfo() ,getVal() ,getTs().toString() ,getNote());
    }

    public int describeContents() {
        return 0;
    }

    public void writeToParcel(Parcel out, int flags) {
        out.writeInt(getId());

        out.writeInt(null == getUsr() ? AppGobalDef.INVALID_ID : getUsr().getId());
        out.writeInt(null == getBudget() ? AppGobalDef.INVALID_ID : getBudget().get_id());

        out.writeString(getInfo());
        out.writeString(getNote());
        out.writeString(getVal().toString());
        out.writeString(getTs().toString());

    }

    public static final Parcelable.Creator<PayNoteItem> CREATOR
            = new Parcelable.Creator<PayNoteItem>() {
        public PayNoteItem createFromParcel(Parcel in) {
            return new PayNoteItem(in);
        }

        public PayNoteItem[] newArray(int size) {
            return new PayNoteItem[size];
        }
    };

    private PayNoteItem(Parcel in)   {
        setId(in.readInt());

        int uid = in.readInt();
        if(AppGobalDef.INVALID_ID != uid)   {
            setUsr(AppModel.getUsrUtility().GetUsrById(uid));
        }

        int bid = in.readInt();
        if(AppGobalDef.INVALID_ID != bid)   {
            setBudget(AppModel.getBudgetUtility().GetBudgetById(bid));
        }

        setInfo(in.readString());
        setNote(in.readString());
        setVal(new BigDecimal(in.readString()));

        try {
            setTs(new Timestamp(0));

            SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US);
            Date date = format.parse(in.readString());
            getTs().setTime(date.getTime());
        }
        catch (ParseException ex)
        {
            setTs(new Timestamp(0));
        }
    }

    public BudgetItem getBudget() {
        return budget;
    }

    public void setBudget(BudgetItem bi) {
        this.budget = bi;
    }
}

File: app/src/main/java/wxm/KeepAccount/item/PayNoteItem.kt
package wxm.KeepAccount.item

import com.j256.ormlite.field.DataType
import com.j256.ormlite.field.DatabaseField
import com.j256.ormlite.table.DatabaseTable
import wxm.KeepAccount.define.GlobalDef
import wxm.KeepAccount.define.IPublicClone
import wxm.KeepAccount.utility.toMoneyString

import java.math.BigDecimal
import java.sql.Timestamp

import wxm.androidutil.db.IDBRow
import java.util.*

/**
 * pay record
 * Created by WangXM on 2016/5/3.
 */
@DatabaseTable(tableName = "tbPayNote")
class PayNoteItem(override var tag: Any? = null) : INote, IDBRow<Int>, Cloneable, IPublicClone {
    @DatabaseField(generatedId = true, columnName = "_id", dataType = DataType.INTEGER)
    override var id: Int = GlobalDef.INVALID_ID

    @DatabaseField(columnName = FIELD_USR, foreign = true, foreignColumnName = UsrItem.FIELD_ID, canBeNull = false)
    override var usr: UsrItem? = null

    @DatabaseField(columnName = FIELD_BUDGET, foreign = true, foreignColumnName = BudgetItem.FIELD_ID)
    override var budget: BudgetItem? = null

    @DatabaseField(columnName = "info", canBeNull = false, dataType = DataType.STRING)
    override var info: String = ""

    @DatabaseField(columnName = "note", dataType = DataType.STRING)
    override var note: String? = null

    @DatabaseField(columnName = "val", dataType = DataType.BIG_DECIMAL)
    override var amount: BigDecimal = BigDecimal.ZERO
        set(newAmount) {
            field = newAmount
            valToStr = field.toMoneyString()
        }

    @DatabaseField(columnName = "ts", dataType = DataType.TIME_STAMP)
    override var ts: Timestamp = Timestamp(0)
        set(tsVal) {
            field = tsVal
            tsToStr = field.toString()
        }

    override var images: LinkedList<NoteImageItem> = LinkedList()
        set(value) {
            field.clear()
            field.addAll(value)
        }

    override var valToStr: String = ""
        get() {
            if(field.isEmpty())  field = amount.toMoneyString()
            return field
        }
        private set(value) {
            field = value
        }

    override var tsToStr: String = ""
        get() {
            if(field.isEmpty())  field = ts.toString()
            return field
        }
        private set(value) {
            field = value
        }

    override val isPayNote: Boolean
        get() = true

    override val isIncomeNote: Boolean
        get() = false


    override fun toPayNote(): PayNoteItem {
        return this
    }

    override fun toIncomeNote(): IncomeNoteItem? {
        return null
    }

    override fun toString(): String {
        return String.format(Locale.CHINA, "info : %s, amount : %f, timestamp : %s\nnote : %s", info, amount, ts.toString(), note)
    }

    override fun getID(): Int {
        return id
    }

    override fun setID(integer: Int) {
        id = integer
    }

    override fun clone(): Any {
        val obj = PayNoteItem(tag)

        obj.id = this.id
        this.usr?.let {
            obj.usr = it.publicClone() as UsrItem
        }

        this.budget?.let {
            obj.budget = it.publicClone() as BudgetItem
        }

        obj.amount = this.amount
        obj.info = this.info
        obj.note = this.note
        obj.ts = this.ts

        obj.valToStr = this.valToStr
        obj.tsToStr = this.tsToStr
        obj.images = this.images
        return obj
    }

    override fun publicClone(): Any {
        return clone()
    }

    companion object {
        const val FIELD_USR = "usr_id"
        const val FIELD_BUDGET = "budget_id"
    }
}

